### PR TITLE
Migrating to the AWS SDK for Go V2

### PR DESCRIPTION
**Pull request by mjura**
The AWS SDK v1 (which is used by this project) is moving to maintenance mode July 2024 and will be completely out of support July 2025.

With the long deprecation policy of Rancher we will still have the eks-operator around when the AWS SDK v1 moves out of support.
We need to update the operator and migrate to v2 of the SDK.

Issue: https://github.com/rancher/eks-operator/issues/336

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**
Issue #

**Special notes for your reviewer**:

**Checklist**:

- [ ] squashed commits into logical changes
- [ ] includes documentation
- [ ] adds unit tests
- [ ] adds or updates e2e tests
- [ ] backport needed

### CODE REVIEW

File: controller/internal.go [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+	"github.com/aws/aws-sdk-go-v2/credentials"
+	"github.com/aws/aws-sdk-go-v2/service/cloudformation"
+	eksv1 "github.com/rancher/eks-operator/pkg/apis/eks.cattle.io/v1"
+	"github.com/rancher/eks-operator/pkg/eks/services"
+	"github.com/rancher/eks-operator/utils"
+	wranglerv1 "github.com/rancher/wrangler/v2/pkg/generated/controllers/core/v1"
+)
+
+/*
+func newAWSSession(secretsCache wranglerv1.SecretCache, spec eksv1.EKSClusterConfigSpec) (*session.Session, error) {
```
Comment (alexander-demicev) [inline controller/internal.go:18]: should these commented functions be removed?

File: controller/nodegroup.go [hunk 2]
```diff
@@ -51,10 +53,15 @@ func deleteLaunchTemplate(templateID string, ec2Service services.EC2ServiceInter
 	)
 }
 
-func deleteNodeGroups(config *eksv1.EKSClusterConfig, nodeGroups []eksv1.NodeGroup, eksService services.EKSServiceInterface) (bool, error) {
+func deleteNodeGroups(ctx context.Context, config *eksv1.EKSClusterConfig, nodeGroups []eksv1.NodeGroup, eksService services.EKSServiceInterface) (bool, error) {
 	var waitingForNodegroupDeletion bool
 	for _, ng := range nodeGroups {
-		_, deleteInProgress, err := deleteNodeGroup(config, ng, eksService)
+		_, deleteInProgress, err := deleteNodeGroup(ctx, config, ng, eksService)
+		/*
```
Comment (alexander-demicev) [inline controller/nodegroup.go:60]: why is this commented?

File: pkg/eks/create.go [hunk 3]
```diff
@@ -128,27 +133,44 @@ func CreateStack(opts *CreateStackOptions) (*cloudformation.DescribeStacksOutput
...
-		events, err := opts.CloudFormationService.DescribeStackEvents(&cloudformation.DescribeStackEventsInput{
+		events, err := opts.CloudFormationService.DescribeStackEvents(ctx, &cloudformation.DescribeStackEventsInput{
 			StackName: aws.String(opts.StackName),
 		})
 		if err == nil {
 			for _, event := range events.StackEvents {
 				// guard against nil pointer dereference
-				if event.ResourceStatus == nil || event.LogicalResourceId == nil || event.ResourceStatusReason == nil {
+				/*
+					if event.ResourceStatus == nil || event.LogicalResourceId == nil || event.ResourceStatusReason == nil {
```
Comment (alexander-demicev) [inline pkg/eks/create.go:148]: same question about commented statements

File: pkg/eks/services/mock_services/doc.go [hunk 4]
```diff
@@ -6,3 +6,7 @@ package mock_services
 //go:generate ../../../../bin/mockgen -destination eks_mock.go -package mock_services -source ../eks.go EKSServiceInterface
 //go:generate ../../../../bin/mockgen -destination iam_mock.go -package mock_services -source ../iam.go IAMServiceInterface
```
Comment (alexander-demicev) [inline pkg/eks/services/mock_services/doc.go:7]: are we keeping v1 services for now?

**Review by alexander-demicev — COMMENTED** [review]
Thanks a lot for taking care of this migration! I have couple minor comments

File: pkg/eks/services/mock_services/doc.go (hunk 4 above)
Comment (mjura) [reply pkg/eks/services/mock_services/doc.go]: I have removed staging code

File: pkg/eks/create.go (hunk 3 above)
Comment (mjura) [reply pkg/eks/create.go]: `event.ResourceStatus` is not a pointer any more
I removed staging comments

File: controller/nodegroup.go (hunk 2 above)
Comment (mjura) [reply controller/nodegroup.go]: it was added for testing only, cleanup

**Review by alexander-demicev — APPROVED** [review]

File: controller/eks-cluster-config-handler_test.go [hunk 5]
```diff
@@ -3,11 +3,12 @@ package controller
...
 	"errors"
 
-	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go-v2/aws"
 
-	"github.com/aws/aws-sdk-go/service/cloudformation"
+	"github.com/aws/aws-sdk-go-v2/service/cloudformation"
+	cftypes "github.com/aws/aws-sdk-go-v2/service/cloudformation/types"
 
-	awssdkeks "github.com/aws/aws-sdk-go/service/eks"
```
Comment (furkatgofurov7) [inline controller/eks-cluster-config-handler_test.go:10]: I believe the old alias was easier to follow compared to current change

File: controller/external.go [hunk 6]
```diff
@@ -17,11 +25,189 @@ func StartAWSSessions(secretsCache wranglerv1.SecretCache, spec eksv1.EKSCluster
...
-		code == eks.NodegroupIssueCodeInstanceLimitExceeded ||
-		code == eks.NodegroupIssueCodeInsufficientFreeAddresses ||
-		code == eks.NodegroupIssueCodeClusterUnreachable
+	return code == string(ekstypes.NodegroupIssueCodeAsgInstanceLaunchFailures) ||
+		code == string(ekstypes.NodegroupIssueCodeInstanceLimitExceeded) ||
+		code == string(ekstypes.NodegroupIssueCodeInsufficientFreeAddresses) ||
+		code == string(ekstypes.NodegroupIssueCodeClusterUnreachable)
+}
+
+// buildUpstreamClusterState
```
Comment (furkatgofurov7) [inline controller/external.go:38]: good to have a proper godoc if we have func name already here

File: controller/internal.go (hunk 1 above)
Comment (mjura) [reply controller/internal.go]: removed

File: controller/eks-cluster-config-handler_test.go (hunk 5 above)
Comment (mjura) [reply controller/eks-cluster-config-handler_test.go:10]: added back previous alias

File: controller/external.go (hunk 6 above)
Comment (mjura) [reply controller/external.go]: good idea, fixed

**Review by furkatgofurov7 — APPROVED** [review]

**Pull request merged**